Add Validator infrastructure

CarlSchwan opened the pull request:
Originally this PR tried to fix the validation of URLs, but the scope of this
changed to instead provide a generic interface for validating user input,
configs, ... The theming API was ported to it as example and to fix the
original bug.

The API is based on Symfony Validator component but the implementation
is simplified a lot since we don't support (yet) validating class with annotation
and other more complicated (and useful) stuff.

<s>The filter_var function is unfortunately not perfect and doesn't support
domain with unicode as well as url with underscores. Replace usage with
a regex.

See https://bugs.php.net/search.php?cmd=display&search_for=FILTER_VALIDATE_URL</s>

Closes #27906

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

juliushaertl commented on apps/theming/lib/Controller/ThemingController.php:
```diff
@@ -203,8 +203,8 @@ public function updateStylesheet($setting, $value) {
-		return ((strpos($url, 'http://') === 0 || strpos($url, 'https://') === 0) &&
-			filter_var($url, FILTER_VALIDATE_URL) !== false);
+		// Regex copied from https://stackoverflow.com/a/206087 but without ftp support
+		return preg_match('#(https?://(\S*?\.\S*?|localhost))([\s)\[\]{},;"\':<]|\.\s|$)#i', $url) === 1;
```
I'd suggest to adjust the regex to also allow hostnames without the domain part other than localhost only

```suggestion
		return preg_match('#(https?://(\S*?\.*\S*?))([\s)\[\]{},;"\':<]|\.\s|$)#i', $url) === 1;
```

juliushaertl replied:
Or just to safeguard, we could only use the regex as a fallback if the filter_var doesn't match:

```suggestion
		return filter_var($url, FILTER_VALIDATE_URL) !== false) || preg_match('#(https?://(\S*?\.\S*?|localhost))([\s)\[\]{},;"\':<]|\.\s|$)#i', $url) === 1;
```

CarlSchwan replied:
It turns out there is a idn_to_ascii function that makes it possible to then use filter_var. Fount it in https://github.com/nextcloud/server/issues/27546.

juliushaertl commented:
Mind to squash to a single commit? :)

blizzz commented:
CI is unhappy tho

CarlSchwan commented:
It's a bit weird that the ci report a failure for continuous-integration/drone/pr but in https://drone.nextcloud.com/nextcloud/server/7311 everything is green

juliushaertl commented:
> It's a bit weird that the ci report a failure for continuous-integration/drone/pr but in [drone.nextcloud.com/nextcloud/server/7311](https://drone.nextcloud.com/nextcloud/server/7311) everything is green

What about https://drone.nextcloud.com/nextcloud/server/7311/9/4 :wink: 

```

There were 4 failures:

1) OCA\Theming\Tests\Controller\ThemingControllerTest::testUpdateStylesheetSuccess with data set #1 ('url', 'https://nextcloud.com/aaaaaaa...aaaaaa', 'Saved')
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
         'data' => Array (
-            'message' => 'Saved'
-            'serverCssUrl' => '/nextcloudWebroot/core/css/so...s.scss'
+            'message' => 'The given web address is not a valid URL'
         )
-        'status' => 'success'
+        'status' => 'error'
     )
     'headers' => Array (...)
     'cookies' => Array ()
-    'status' => 200
+    'status' => 400
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\EmptyContentSecurityPolicy Object (...)

/drone/src/apps/theming/tests/Controller/ThemingControllerTest.php:168

2) OCA\Theming\Tests\Controller\ThemingControllerTest::testUpdateStylesheetSuccess with data set #6 ('imprintUrl', 'https://nextcloud.com/aaaaaaa...aaaaaa', 'Saved')
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
         'data' => Array (
-            'message' => 'Saved'
-            'serverCssUrl' => '/nextcloudWebroot/core/css/so...s.scss'
+            'message' => 'The given legal notice address is not a valid URL'
         )
-        'status' => 'success'
+        'status' => 'error'
     )
     'headers' => Array (...)
     'cookies' => Array ()
-    'status' => 200
+    'status' => 400
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\EmptyContentSecurityPolicy Object (...)

/drone/src/apps/theming/tests/Controller/ThemingControllerTest.php:168

3) OCA\Theming\Tests\Controller\ThemingControllerTest::testUpdateStylesheetSuccess with data set #7 ('privacyUrl', 'https://nextcloud.com/aaaaaaa...aaaaaa', 'Saved')
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
         'data' => Array (
-            'message' => 'Saved'
-            'serverCssUrl' => '/nextcloudWebroot/core/css/so...s.scss'
+            'message' => 'The given privacy policy address is not a valid URL'
         )
-        'status' => 'success'
+        'status' => 'error'
     )
     'headers' => Array (...)
     'cookies' => Array ()
-    'status' => 200
+    'status' => 400
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\EmptyContentSecurityPolicy Object (...)

/drone/src/apps/theming/tests/Controller/ThemingControllerTest.php:168

4) OCA\Theming\Tests\Controller\ThemingControllerTest::testUpdateStylesheetError with data set #1 ('url', 'http://example.com/aaaaaaaaaa...aaaaaa', 'The given web address is too long')
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
         'data' => Array (
-            'message' => 'The given web address is too long'
+            'message' => 'The given web address is not a valid URL'
         )
         'status' => 'error'
     )

/drone/src/apps/theming/tests/Controller/ThemingControllerTest.php:217

--
```

CarlSchwan commented:
It seems `idn_to_ascii` is kinda evil too and just abort when the URL is too long :(

A better solution would be either to just copy the UrlValidator::validate from symfony/validator or just depends on symfony/validator and use it for all kind of validations in Nextcloud (URL, email, date ...). For reference, this is the method: https://github.com/symfony/validator/blob/14337bdf9e2e0b2e3385c9e90f13325f0c95a4f9/Constraints/UrlValidator.php#L24

tcitworld commented:
> It seems idn_to_ascii is kinda evil too and just abort when the URL is too long :(

That's probably because it's made to handle domains, not full URIs. You could extract the hostname with `parse_url`, convert it, then `substr_replace` it in the URL.

But using the regex from SF Validator makes sense too.

eneiluj commented on lib/public/Util.php:
```diff
@@ -513,4 +513,36 @@ public static function needUpgrade() {
+			(?:\# (?:[\pL\pN\-._\~!$&\'()*+,;=:@/?]|%%[0-9A-Fa-f]{2})* )?       # a fragment (optional)
+		$~ixu';
+		$pattern = sprintf($pattern, implode('|', $protocols));
+		return preg_match($pattern, $url) !== false;
```
`isValidUrl` will always return true unless an error occurs in `preg_match`.
```suggestion
		return preg_match($pattern, $url) === 1;
```
https://www.php.net/manual/en/function.preg-match.php#refsect1-function.preg-match-returnvalues
Or maybe it is intentional. Is it?

eneiluj commented on apps/oauth2/lib/Controller/SettingsController.php:
```diff
@@ -81,7 +81,7 @@ public function __construct(string $appName,
 	public function addClient(string $name,
 							  string $redirectUri): JSONResponse {
-		if (filter_var($redirectUri, FILTER_VALIDATE_URL) === false) {
+		if (!\OCP\Util::isValidUrl($redirectUri)) {
```
Sorry, small detail, `OCP\Util` could be imported: `use OCP\Util;`
```suggestion
		if (!Util::isValidUrl($redirectUri)) {
```
like it's done in `apps/theming/lib/Controller/ThemingController.php`

eneiluj commented on apps/theming/lib/ThemingDefaults.php:
```diff
@@ -203,7 +203,7 @@ public function getShortFooter() {
 		foreach ($links as $link) {
 			if ($link['url'] !== ''
-				&& filter_var($link['url'], FILTER_VALIDATE_URL)
+				&& \OCP\Util::isValidUrl($link['url'])
```
Same here :grin:

CarlSchwan replied:
There was a Util class in the same file and it collided so I used an alias OCPUtil

eneiluj commented on lib/private/Setup.php:
```diff
@@ -474,7 +474,7 @@ private static function findWebRoot(SystemConfig $config): string {
 				throw new InvalidArgumentException('overwrite.cli.url is empty');
 			}
-			if (!filter_var($webRoot, FILTER_VALIDATE_URL)) {
+			if (!\OCP\Util::isValidUrl($webRoot)) {
```
Here

julien-nc reviewed (dismissed):
Apart from that, code looks good. Didn't try it though.

szaimen commented:
What is missing here?

PVince81 commented:
@CarlSchwan are you able to fix the tests ?

szaimen commented:
/backport to stable23

szaimen commented:
/backport to stable22

szaimen committed:
Fix IDN domain name not being allowed

The filter_var function is unfortunately not perfect and doesn't support
domain with unicode as well as url with underscores. Replace usage with
a regex from Symfony validator.

See https://bugs.php.net/search.php?cmd=display&search_for=FILTER_VALIDATE_URL

Closes #27906

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

szaimen commented:
I fixed the conflicts and rebased

szaimen commented:
@CarlSchwan please fix the tests if possible :)

CarlSchwan committed:
Add an validator service with various constraints

The API is heavily inspired by Symfony validator while having a much
simpler implementation since this doesn't do object validation with
annotation.

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Add missing copyright

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan commented on tests/lib/Validator/ValidatorTest.php:
```diff
@@ -0,0 +1,123 @@
+<?php
+/**
+ * Copyright (c) 2012 Bernhard Posselt <dev@bernhard-posselt.com>
```
copy pasta :(

tcitworld commented on lib/private/Validator/IConstraintValidator.php:
```diff
@@ -0,0 +1,33 @@
+ *
+ */
+
+namespace OC\Validator;
```
Not in OCP ?

CarlSchwan replied:
For now, since I'm trying to keep the interface surface small to make it possible to easily extend it in the future. I'm for example not completely sure that it is helpful to return an array of Violations, as for the moment all the validator are returning only one or zero violation.

If you have some usecase in mind that would require to have that in OCP, I'm happy to reconsider that (and in that case, I would also probably merge the Constraint and Validator).

CarlSchwan replied:
For now, since I'm trying to keep the interface surface small to make it possible to easily extend it in the future. I'm for example not completely sure that it is helpful to return an array of Violations, as for the moment all the validator are returning only one or zero violation.

If you have some usecase in mind that would require to have that in OCP, I'm happy to reconsider that (and in that case, I would also probably merge the Constraint and Validator).

come-nc commented on lib/public/Validator/Violation.php:
```diff
@@ -0,0 +1,65 @@
+/**
+ * This object represents a constraint violation when validating a value.
+ */
+class Violation {
```
Maybe this should extends Error or Exception, to be able to get traces for debug, not sure

CarlSchwan replied:
> What is not clear to me:
> 
>     * if "The API is based on Symfony Validator", why not use symfony validator instead of coding our own?

Because the Symfony Validator is really heavy in term of dependencies (e.g. requires HTTP foundation and many other core components of Symfony). It is not that reusable outside of Symfony :( 

>     * Why is there for each type (url, csscolor…) both a validator and a constraint, couldn’t that be only one class? Seems overly complicated. Is there sometimes several validators for the same constraint, or several constraints for the same validator?

For two reasons:

* Keep the validator itself as a private API that apps shouldn't use directly
* It would allow writing multiple Constraints that internally use the same Validator

But I agree it's a bit complicated and if we make `validate` a public API, then we could merge the two

come-nc reviewed (commented):
What is not clear to me:
- if "The API is based on Symfony Validator", why not use symfony validator instead of coding our own?
- Why is there for each type (url, csscolor…) both a validator and a constraint, couldn’t that be only one class? Seems overly complicated. Is there sometimes several validators for the same constraint, or several constraints for the same validator?

CarlSchwan committed:
Apply suggestions from review

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

szaimen committed:
add validator to setup check

Signed-off-by: szaimen <szaimen@e.mail.de>

CarlSchwan committed:
Merge constraint and validator

This simplifies the code a bit and make it possible for apps to add
their own validators

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
fixup! Merge constraint and validator

CarlSchwan committed:
fixup! Merge constraint and validator

CarlSchwan committed:
fixup! Merge constraint and validator

CarlSchwan committed:
fixup! Merge constraint and validator

artonge commented on apps/settings/lib/Controller/CheckSetupController.php:
```diff
@@ -618,7 +624,10 @@ protected function getSuggestedOverwriteCliURL(): string {
 		// Check correctness by checking if it is a valid URL
-		if (filter_var($currentOverwriteCliUrl, FILTER_VALIDATE_URL)) {
+		if ($this->validator->isValid($currentOverwriteCliUrl, [
+			new NotBlank(),
```
Shouldn't Url already check for NotBlank?

szaimen commented:
conflicts :/
